### PR TITLE
[W.I.P.] Add more granular Publish tests

### PR DESCRIPTION
**Pull request by IshanMi**
## Description
At the moment we only test the output of our DP system- i.e. the output of calling `.publish()` on a DP tensor. This results in  the `publish` method being a black box where things may break and go undetected. Examples of this include `jnp.take` changing behaviour and not raising Out of Bounds/IndexErrors, or privacy budgets not spending properly.

This PR with @curt-mitch aims to test the internals of the publish method, and ensure that it fails gracefully and with some trace of the associated error. This will hopefully give us more confidence about the reliability of the DP system and the publish method.

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

**Comment by iamtrask** [comment]
Very excited about this PR - well framed and timely. Also great job for creating a WIP ahead of time with a great description. This greatly helps with visibility on what features/changes are coming.

**Comment by rasswanth-s** [comment]
@IshanMi , would this PR functionality be part of the 0.7 release or for future release ?

**Comment by rasswanth-s** [comment]
@IshanMi , could we close the PR, as it is related to 0.7, or do we have some changes in the PR, we would be using later ?
